Adding a custom form validation topic (#6886)
magento/devdocs: src/guides/v2.3/frontend-dev-guide/validations/custom-form-validation.md

---
group: frontend-developer-guide
title: Custom form validation
contributor_name: Atwix
contributor_link: https://www.atwix.com/
---

This tutorial shows you how to validate your own form, using `mage/validation` library, before submitting it to the server.

> This gives the possibility to customer to fix all the invalid data straight away, which improves the good user experience.

Reviewer: As a rule, we don't need to improve a good experience. Unless we are going to make it perfect 🙂

```suggestion
This gives the possibility to customer to fix all the invalid data straight away, which improves the user experience and user interface accessibility.
```

## Initiating validation

There are several ways to initiate the form validation for a form as described in [Initializing JavaScript]({{ page.baseurl }}/javascript-dev-guide/javascript/js_init.html) topic.

### Using the `data-mage-init` attribute

```html
<form id="my-form" data-mage-init='{"validation": {}}'>
    ...
</form>
```

### Using the `<script type="text/x-magento-init">` tag

```html
<script type="text/x-magento-init">
    {
        "#my-form": {
            "validation": {}
        }
    }
</script>
```

## Form validation rules

All the available Magento validation rules may be found in [validation/rules.js]({{ site.mage2bloburl }}/{{page.guide_version}}/app/code/Magento/Ui/view/base/web/js/lib/validation/rules.js)

Additionally, you may also use any available [jQuery validation rule](https://jqueryvalidation.org/documentation/#link-list-of-built-in-validation-methods).

## Defining validation rules

There are couple of ways to define the validation rules for a form field.

### As `data-validate` attribute

```html
<input id="field-1" ... data-validate='{"required":true}'/>
```

### As attribute

```html

> <input id="field-1" ... required="true"'/>

Reviewer: ```suggestion
<input id="field-1" ... required="true"/>
```

```

### As class name

```html

> <input id="field-1" ... class="input-text required-entry"'/>

Reviewer: ```suggestion
<input id="field-1" ... class="input-text required-entry"/>
```
